### PR TITLE
Remove copy from SUNDIALS::KINSOL

### PR DESCRIPTION
**Pull request by peterrum**

### CODE REVIEW

File: include/deal.II/sundials/n_vector.templates.h [hunk 1]
```diff
@@ -168,6 +168,14 @@ namespace SUNDIALS
       realtype
       dot_product(N_Vector x, N_Vector y);
 
+      template <typename VectorType>
+      realtype
+      weigted_l2_norm(N_Vector x, N_Vector y);
```
Comment (luca-heltai) [inline include/deal.II/sundials/n_vector.templates.h:173]:
```suggestion
      weighted_l2_norm(N_Vector x, N_Vector y);
```

File: include/deal.II/sundials/n_vector.templates.h [hunk 2]
```diff
@@ -569,6 +577,28 @@ SUNDIALS::internal::NVectorOperations::dot_product(N_Vector x, N_Vector y)
 
 
 
+template <typename VectorType>
+realtype
+SUNDIALS::internal::NVectorOperations::weigted_l2_norm(N_Vector x, N_Vector w)
```
Comment (luca-heltai) [inline include/deal.II/sundials/n_vector.templates.h:582]:
```suggestion
SUNDIALS::internal::NVectorOperations::weigthed_l2_norm(N_Vector x, N_Vector w)
```

File: include/deal.II/sundials/n_vector.templates.h [hunk 3]
```diff
@@ -911,9 +941,10 @@ SUNDIALS::internal::create_empty_nvector()
   v->ops->nvmaxnorm   = NVectorOperations::max_norm<VectorType>;
   v->ops->nvwrmsnorm  = NVectorOperations::weighted_rms_norm<VectorType>;
   //  v->ops->nvwrmsnormmask = undef;
-  v->ops->nvmin = NVectorOperations::min_element<VectorType>;
-  //  v->ops->nvwl2norm      = undef;
-  //  v->ops->nvl1norm       = undef;
+  v->ops->nvmin     = NVectorOperations::min_element<VectorType>;
+  v->ops->nvwl2norm = NVectorOperations::weigted_l2_norm<VectorType>;
```
Comment (luca-heltai) [inline include/deal.II/sundials/n_vector.templates.h:945]:
```suggestion
  v->ops->nvwl2norm = NVectorOperations::weighted_l2_norm<VectorType>;
```

File: source/sundials/kinsol.cc [hunk 4]
```diff
@@ -35,6 +35,7 @@
 #  endif
 
 #  include <deal.II/sundials/copy.h>
```
Comment (luca-heltai) [inline source/sundials/kinsol.cc:37]: Remove this header if it is no longer used.

File: source/sundials/kinsol.cc [hunk 5]
```diff
@@ -137,15 +138,9 @@ namespace SUNDIALS
...
-      GrowingVectorMemory<VectorType> mem;
 
-      typename VectorMemory<VectorType>::Pointer src_yy(mem);
-      solver.reinit_vector(*src_yy);
-
-      typename VectorMemory<VectorType>::Pointer dst_FF(mem);
-      solver.reinit_vector(*dst_FF);
-
-      copy(*src_yy, yy);
+      auto *src_yy = internal::unwrap_nvector<VectorType>(yy);
```
Comment (luca-heltai) [inline source/sundials/kinsol.cc:142]:
```suggestion
      auto *src_yy = internal::unwrap_nvector_const<VectorType>(yy);
```

File: source/sundials/kinsol.cc [hunk 6]
```diff
@@ -169,16 +162,10 @@ namespace SUNDIALS
...
 
-      typename VectorMemory<VectorType>::Pointer src_ycur(mem);
-      solver.reinit_vector(*src_ycur);
-
-      typename VectorMemory<VectorType>::Pointer src_fcur(mem);
-      solver.reinit_vector(*src_fcur);
-
-      copy(*src_ycur, kinsol_mem->kin_uu);
-      copy(*src_fcur, kinsol_mem->kin_fval);
+      auto *src_ycur = internal::unwrap_nvector<VectorType>(kinsol_mem->kin_uu);
```
Comment (luca-heltai) [inline source/sundials/kinsol.cc:166]:
```suggestion
      auto *src_ycur = internal::unwrap_nvector_const<VectorType>(kinsol_mem->kin_uu);
```

File: source/sundials/kinsol.cc [hunk 7]
```diff
@@ -169,16 +162,10 @@ namespace SUNDIALS
...
+      auto *src_fcur =
+        internal::unwrap_nvector<VectorType>(kinsol_mem->kin_fval);
```
Comment (luca-heltai) [inline source/sundials/kinsol.cc:168]:
```suggestion
      auto *src_fcur =
        internal::unwrap_nvector_const<VectorType>(kinsol_mem->kin_fval);
```

File: source/sundials/kinsol.cc [hunk 8]
```diff
@@ -196,27 +183,14 @@ namespace SUNDIALS
...
-      solver.reinit_vector(*src);
-
-      typename VectorMemory<VectorType>::Pointer dst(mem);
-      solver.reinit_vector(*dst);
-
-      copy(*src, b);
+      auto *src_ycur = internal::unwrap_nvector<VectorType>(kinsol_mem->kin_uu);
+      auto *src_fcur =
+        internal::unwrap_nvector<VectorType>(kinsol_mem->kin_fval);
+      auto *src = internal::unwrap_nvector<VectorType>(b);
```
Comment (luca-heltai) [inline source/sundials/kinsol.cc:190]:
```suggestion
      auto *src = internal::unwrap_nvector_const<VectorType>(b);
```

File: source/sundials/kinsol.cc [hunk 9]
```diff
@@ -196,27 +183,14 @@ namespace SUNDIALS
...
-      typename VectorMemory<VectorType>::Pointer src(mem);
-      solver.reinit_vector(*src);
-
-      typename VectorMemory<VectorType>::Pointer dst(mem);
-      solver.reinit_vector(*dst);
-
-      copy(*src, b);
+      auto *src_ycur = internal::unwrap_nvector<VectorType>(kinsol_mem->kin_uu);
+      auto *src_fcur =
+        internal::unwrap_nvector<VectorType>(kinsol_mem->kin_fval);
```
Comment (luca-heltai) [inline source/sundials/kinsol.cc:189]:
```suggestion
        internal::unwrap_nvector_const<VectorType>(kinsol_mem->kin_fval);
```

File: source/sundials/kinsol.cc [hunk 10]
```diff
@@ -309,13 +265,9 @@ namespace SUNDIALS
...
           typename VectorMemory<VectorType>::Pointer src_ycur(mem);
           typename VectorMemory<VectorType>::Pointer src_fcur(mem);
-          typename VectorMemory<VectorType>::Pointer src_b(mem);
-          typename VectorMemory<VectorType>::Pointer dst_x(mem);
-
-          solver.reinit_vector(*src_b);
-          solver.reinit_vector(*dst_x);
 
-          copy(*src_b, b);
+          auto *src_b = internal::unwrap_nvector<VectorType>(b);
```
Comment (luca-heltai) [inline source/sundials/kinsol.cc:269]:
```suggestion
          auto *src_b = internal::unwrap_nvector_const<VectorType>(b);
```

**Review by luca-heltai — APPROVED** [review]

**Comment by luca-heltai** [comment]
Please address the minor changes above. Otherwise, it seems good to me!

**Review by kronbichler — APPROVED** [review]

**Pull request merged**